google/cloud/internal/external_account_token_source_aws.cc

```diff
@@ -95,6 +115,94 @@ StatusOr<ExternalAccountTokenSourceAwsInfo> ParseExternalAccountTokenSourceAws(
+    HttpClientFactory const& client_factory, Options const& opts,
+    internal::ErrorContext const& /*ec*/) {
+  if (info.imdsv2_session_token_url.empty()) return std::string{};
+  auto client = client_factory(opts);
+  auto request =
+      rest_internal::RestRequest()
+          .SetPath(std::move(info.imdsv2_session_token_url))
+          .AddHeader(kMetadataTokenTtlHeader,
+                     std::to_string(kDefaultMetadataTokenTtl.count()));
+  auto response = client->Put(request, {});
+  if (!response) return std::move(response).status();
+  if (IsHttpError(**response)) return AsStatus(std::move(**response));
+  return rest_internal::ReadAll(std::move(**response).ExtractPayload());
+}
+
+StatusOr<std::string> FetchRegion(ExternalAccountTokenSourceAwsInfo const& info,
+                                  std::string const& metadata_token,
+                                  HttpClientFactory const& cf,
+                                  Options const& opts,
+                                  internal::ErrorContext const& ec) {
+  for (auto const* name : {"AWS_REGION", "AWS_DEFAULT_REGION"}) {
+    auto env = internal::GetEnv(name);
+    if (env.has_value()) return std::move(*env);
+  }
+
+  auto payload = GetMetadata(info.region_url, metadata_token, cf, opts);
+  if (!payload) return std::move(payload).status();
+  if (payload->empty()) {
+    return InvalidArgumentError(
+        absl::StrCat("invalid (empty) region returned from ", info.region_url),
+        GCP_ERROR_INFO().WithContext(ec));
+  }
+  // The metadata service returns an availability zone, we must remove the last
+  // character to return the region.
+  payload->pop_back();
+  return *std::move(payload);
+}
+
+StatusOr<ExternalAccountTokenSourceAwsSecrets> FetchSecrets(
+    ExternalAccountTokenSourceAwsInfo const& info,
+    std::string const& metadata_token, HttpClientFactory const& cf,
+    Options const& opts, internal::ErrorContext const& ec) {
+  auto access_key_id_env = internal::GetEnv("AWS_ACCESS_KEY_ID");
+  auto secret_access_key_env = internal::GetEnv("AWS_SECRET_ACCESS_KEY");
+  auto session_token_env = internal::GetEnv("AWS_SESSION_TOKEN");
+  if (access_key_id_env.has_value() && secret_access_key_env.has_value()) {
+    return ExternalAccountTokenSourceAwsSecrets{
+        /*access_key_id=*/std::move(*access_key_id_env),
+        /*secret_access_key=*/std::move(*secret_access_key_env),
+        /*session_token=*/session_token_env.value_or("")};
+  }
+
+  // This code fetches the security credentials from the metadata services in an
+  // AWS EC2 instance, i.e., a VM. The requests and responses are documented in:
+  //  https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+  auto role = GetMetadata(info.url, metadata_token, cf, opts);
+  if (!role) return std::move(role).status();
+  auto path = info.url;
+  if (path.back() != '/') path.push_back('/');
```

Comment:
Do we need check `path.empty()`?

Reply:
I am going to leave this as-is for now. This will only be used with an `info` initialized with `ParseExternalAccountTokenSourceAws()` and I will be adding some code to prevent empty `url` values to that function.